azkaban-common/src/main/java/azkaban/project/FlowLoader.java

```diff
@@ -0,0 +1,139 @@
+/*
+* Copyright 2017 LinkedIn Corp.
+*
+* Licensed under the Apache License, Version 2.0 (the “License”); you may not
+* use this file except in compliance with the License. You may obtain a copy of
+* the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an “AS IS” BASIS, WITHOUT
+* WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+* License for the specific language governing permissions and limitations under
+* the License.
+*/
+
+package azkaban.project;
+
+import azkaban.flow.CommonJobProperties;
+import azkaban.flow.Flow;
+import azkaban.project.validator.ValidationReport;
+import azkaban.utils.Props;
+import java.io.File;
+import java.io.FileFilter;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Loads flows from project directory.
+ */
+public abstract class FlowLoader {
```

Comment:
Have you considered using the composition instead of inheritance see the dev guide.

Reply:
I've thought about using composition. However, I don't think it will benefit from using composition in this case. Flowloader does not actually have any major implementation of loading project flows. So even if we add it as a reference inside DirectoryFlowLoader and DirectoryYamlFlowLoader, it doesn't make any difference. I think the better way to do it would be to change FlowLoader to an interface with one loadProject method inside. Then DirectoryFlowLoader and DirectoryYamlFlowLoader can extend the interface and implement their own loadProject method. We can move some common method like addEmailPropsToFlow to some util class. What do you think of the new approach? We can discuss it offline.